nimbus/src/enrollment.rs

```diff
@@ -85,90 +365,156 @@ pub fn get_enrollments(db: &Database) -> Result<Vec<EnrolledExperiment>> {
+        updated_experiments: &[Experiment],
+        existing_enrollments: &[ExperimentEnrollment],
+    ) -> Result<(Vec<ExperimentEnrollment>, Vec<EnrollmentChangeEvent>)> {
+        let mut updated_enrollments = vec![];
+        let mut enrollment_events = vec![];
+        let existing_experiments = map_experiments(&existing_experiments);
+        let updated_experiments = map_experiments(&updated_experiments);
+        let existing_enrollments = map_enrollments(&existing_enrollments);
+
+        let mut all_slugs = HashSet::with_capacity(existing_experiments.len());
+        all_slugs.extend(existing_experiments.keys());
+        all_slugs.extend(updated_experiments.keys());
+        all_slugs.extend(existing_enrollments.keys());
+
+        for slug in all_slugs {
+            let updated_enrollment = self.evolve_enrollment(
+                is_user_participating,
+                existing_experiments.get(slug).copied(),
+                updated_experiments.get(slug).copied(),
+                existing_enrollments.get(slug).copied(),
+                &mut enrollment_events,
+            )?;
+            if let Some(enrollment) = updated_enrollment {
+                updated_enrollments.push(enrollment);
+            }
+        }
+
+        Ok((updated_enrollments, enrollment_events))
+    }
+
+    fn evolve_enrollment(
+        &self,
+        is_user_participating: bool,
+        existing_experiment: Option<&Experiment>,
+        updated_experiment: Option<&Experiment>,
+        existing_enrollment: Option<&ExperimentEnrollment>,
+        enrollment_events: &mut Vec<EnrollmentChangeEvent>,
+    ) -> Result<Option<ExperimentEnrollment>> {
+        Ok(
+            match (existing_experiment, updated_experiment, existing_enrollment) {
+                // New experiment.
+                (None, Some(experiment), None) => Some(ExperimentEnrollment::from_new_experiment(
+                    is_user_participating,
+                    self.nimbus_id,
+                    self.available_randomization_units,
+                    self.app_context,
+                    experiment,
+                    enrollment_events,
+                )?),
+                // Experiment deleted remotely.
+                (Some(_), None, Some(enrollment)) => {
+                    enrollment.on_experiment_discarded(enrollment_events)
+                }
+                // Known experiment.
+                (Some(_), Some(updated_experiment), Some(existing_enrollment)) => {
+                    existing_enrollment.on_experiment_updated(
+                        is_user_participating,
+                        self.nimbus_id,
+                        self.available_randomization_units,
+                        self.app_context,
+                        updated_experiment,
+                        enrollment_events,
+                    )?
+                }
+                (None, None, Some(_)) => unreachable!("Orphan enrollment."),
```

Comment:
I'm not convinced that this state is unreachable, as we could have a `WasEnrolled` state that has been kept around after an experiment has ended. At the point in time when the experiment ends, we'd make a transition like:

```
(Some(_), None, Some(Enrolled(...))) => Some(WasEnrolled(...))
```

And then on some future update, we'll slurp this `WasEnrolled` state out of the db and process another transition like:

```
(None, None, Some(WasEnrolled(...))) => panic!
```

IIUC, this seems like the place where we want to garbage-collect old `WasEnrolled` states once sufficient time has passed. So instead of panicing, this should inspect the timestamp and either return `None` or `Some(WasEnrolled(...))`.

Reply:
Good catch, and that answers https://github.com/mozilla/nimbus-sdk/pull/69#discussion_r532981366.